src/data-source/collection-browser-data-source.ts

```diff
@@ -0,0 +1,1189 @@
+  addPage(pageNum: number, pageTiles: TileModel[]): void;
+
+  /**
+   * Returns the given page of tile models from the data source.
+   * @param pageNum Which page number to get (indexed starting from 1)
+   */
+  getPage(pageNum: number): TileModel[];
+
+  /**
+   * Returns the full set of paged tile models stored in this data source.
+   */
+  getAllPages(): Record<string, TileModel[]>;
+
+  /**
+   * Whether the data source contains any tiles for the given page number.
+   * @param pageNum Which page number to query (indexed starting from 1)
+   */
+  hasPage(pageNum: number): boolean;
+
+  /**
+   * Returns the single tile model appearing at the given index in the
+   * data source, with respect to the current page size. Returns `undefined` if
+   * the corresponding page is not present on the data source or if it does not
+   * contain a tile model at the corresponding index.
+   * @param index The 0-based index (within the full data source) of the tile to get
+   */
+  getTileModelAt(index: number): TileModel | undefined;
+
+  /**
+   * Requests that the data source fire a backend request for the given page of results.
+   * @param pageNum Which page number to fetch results for
+   * @param numInitialPages How many pages should be batched together on an initial fetch
+   */
+  fetchPage(pageNum: number, numInitialPages?: number): Promise<void>;
+
+  /**
+   * Requests that the data source update its prefix bucket result counts for the given
+   * type of prefix filter.
+   * @param filterType Which prefixable field to update the buckets for (e.g., title/creator)
+   */
+  updatePrefixFilterCounts(filterType: PrefixFilterType): Promise<void>;
+
+  /**
+   * Changes the page size used by the data source, discarding any previously-fetched pages.
+   *
+   * **Note: this operation will reset any data stored in the data source!**
+   * @param pageSize
+   */
+  setPageSize(pageSize: number): void;
+
+  /**
+   * Resets the data source to its empty state, with no result pages, aggregations, etc.
+   */
+  reset(): void;
+
+  /**
+   * Notifies the data source that a query change has occurred, which may trigger a data
+   * reset & new fetches.
+   */
+  handleQueryChange(): void;
+
+  /**
+   * Applies the given map function to all of the tile models in every page of the data
+   * source.
+   * @param callback A callback function to apply on each tile model, as with Array.map
+   */
+  map(
+    callback: (model: TileModel, index: number, array: TileModel[]) => TileModel
+  ): void;
+
+  /**
+   * Checks every tile's management checkbox
+   */
+  checkAllTiles(): void;
+
+  /**
+   * Unchecks every tile's management checkbox
+   */
+  uncheckAllTiles(): void;
+
+  /**
+   * Removes all tile models that are currently checked & adjusts the paging
+   * of the data source to account for any new gaps in the data.
+   */
+  removeCheckedTiles(): void;
+}
+
+export class CollectionBrowserDataSource
+  implements CollectionBrowserDataSourceInterface
```

Comment:
Q: what is the overall responsibility for this class - is it to normalize the results data to create view model? can we add 1-2 lines on its usage?

Reply:
also i didn't see a discreet unit test for this class - do the current tests catch these updates?

Reply:
I'll add some additional comments around the responsibility & usage, thanks! But yeah, it abstracts out a cleaner model for the server-provided data and the logic determining when that data needs to be modified. There's likely still room for a future pass that improves on the abstraction even further, but I'm wary of any more scope bloat on this ticket.

Many of the existing collection browser unit tests do naturally test all of these data source methods (since CB's own functionality is pretty limited without the data source attached). That said, I can still go in and organize those better. There are probably a few gaps that need filling.